test/gtest/ucp/test_ucp_tag_cancel.cc

```diff
@@ -9,29 +9,84 @@
 
 #include <common/test_helpers.h>
 
+#if _OPENMP
+#include "omp.h"
+#endif
 
 class test_ucp_tag_cancel : public test_ucp_tag {
 public:
     using test_ucp_tag::get_ctx_params;
+
+    static std::vector<ucp_test_param> enum_test_params(const ucp_params_t& ctx_params,
+                                                        const ucp_worker_params_t& worker_params,
+                                                        const std::string& name,
+                                                        const std::string& test_case_name,
+                                                        const std::string& tls)
+    {
+        std::vector<ucp_test_param> result;
+
+        generate_test_params_variant(ctx_params, worker_params, name, test_case_name, tls,
+                                     RECV_REQ_INTERNAL, result, MULTI_THREAD_CONTEXT);
+        generate_test_params_variant(ctx_params, worker_params, name, test_case_name, tls,
+                                     RECV_REQ_EXTERNAL, result, MULTI_THREAD_CONTEXT);
+        generate_test_params_variant(ctx_params, worker_params, name, test_case_name, tls,
+                                     RECV_REQ_INTERNAL, result, MULTI_THREAD_WORKER);
+        generate_test_params_variant(ctx_params, worker_params, name, test_case_name, tls,
+                                     RECV_REQ_EXTERNAL, result, MULTI_THREAD_WORKER);
+        return result;
+    }
 };
 
 UCS_TEST_P(test_ucp_tag_cancel, cancel_exp) {
-    uint64_t recv_data = 0;
-    request *req;
-
-    req = recv_nb(&recv_data, sizeof(recv_data), DATATYPE, 1, 1);
-    if (UCS_PTR_IS_ERR(req)) {
-        ASSERT_UCS_OK(UCS_PTR_STATUS(req));
-    } else if (req == NULL) {
-        UCS_TEST_ABORT("ucp_tag_recv_nb returned NULL");
+    uint64_t recv_data[MT_TEST_NUM_THREADS] GTEST_ATTRIBUTE_UNUSED_;
+
+    for (int i = 0; i < MT_TEST_NUM_THREADS; i++) {
+        recv_data[i] = 0;
     }
 
-    ucp_request_cancel(receiver().worker(), req);
-    wait(req);
+    if (GetParam().thread_type == SINGLE_THREAD) {
+        request *req;
```

Comment:
this code can be extracted to a function and used for both single/multi thread tests (same comment for other places)